zcash_test_vectors/zip_0244_zsa.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+import sys; assert sys.version_info[0] >= 3, "Python 3 required."
+
+from hashlib import blake2b
+import struct
+
+from .transaction_zsa import (
+    TransactionZSA, IssueActionDescription,
+)
+from .output import render_args, render_tv
+from .rand import Rand
+from .zip_0143 import (
+    SIGHASH_ALL,
+    SIGHASH_ANYONECANPAY,
+    SIGHASH_NONE,
+    SIGHASH_SINGLE,
+)
+
+# - Burn
+def orchard_zsa_burn_digest(tx: TransactionZSA):
+    digest = blake2b(digest_size=32, person=b'ZTxIdOrcBurnHash')
+
+    if len(tx.vAssetBurnOrchardZSA) > 0:
+        for desc in tx.vAssetBurnOrchardZSA:
+            digest.update(bytes(desc.assetBase))
+            digest.update(struct.pack('<Q', desc.valueBurn))
+
+    return digest.digest()
+
+def issuance_digest(tx):
+    digest = blake2b(digest_size=32, person=b'ZTxIdSAIssueHash')
+
+    if len(tx.vIssueActions) > 0:
+        digest.update(issue_actions_digest(tx))
+        digest.update(tx.ik)
+
+    return digest.digest()
+
+def issuance_auth_digest(tx):
+    digest = blake2b(digest_size=32, person=b'ZTxAuthZSAOrHash')
+    if len(tx.vIssueActions) > 0:
+        digest.update(tx.issueAuthSig)
+    return digest.digest()
+
+def issue_actions_digest(tx):
+    digest = blake2b(digest_size=32, person=b'ZTxIdIssuActHash')
+
+    for action in tx.vIssueActions:
+        digest.update(issue_notes_digest(action))
+        digest.update(action.asset_desc)
+        digest.update(struct.pack('<B', action.flagsIssuance))
+
+    return digest.digest()
+
+def issue_notes_digest(action: IssueActionDescription):
+    digest = blake2b(digest_size=32, person=b'ZTxIdIAcNoteHash')
+
+    for note in action.vNotes:
+        digest.update(bytes(note.recipient))
+        digest.update(struct.pack('<Q', note.value))
+        digest.update(bytes(note.assetBase))
+        digest.update(bytes(note.rho))
+        digest.update(note.rseed)
+
+    return digest.digest()
+
+def main():
+    from .zip_0244 import rand_gen, populate_test_vector, generate_test_vectors
```

Comment:
move to top

Reply:
It doesn't work, because there is a circular import: 
- These three functions (`rand_gen`, `populate_test_vector`, `generate_test_vectors`) are defined in `zip_0244.py` and used in `zip_0244_zsa.py` (now `orchard_zsa/digests.py`)
- The functions `orchard_zsa_burn_digest`, `issuance_digest`, `issuance_auth_digest` are defined in `zip_0244_zsa.py` (now `orchard_zsa/digests.py`) and used in `zip_0244.py`

Importing only in `main` delays the import in a way that avoids this being an issue.

Reply:
One alternative idea I thought of was moving the functions in the first bullet above into a separate file. If we want to do similar code deduplication in the other files that are present for both `orchard` and orchard_zsa` then we can do it in these new files.

The `rand_gen` function should be reusable between all such files, but there are differences in the other two functions so there will have to be a version of them for every pair of files (eg. `orchard/key_components` vs `orchard_zsa/key_components`, or `orchard/note_encryption` vs `orchard_zsa/note_encryption`).

Reply:
Rearranged the files in a way that solves this problem, and moves the import up to the top.